dask/array/rechunk.py

```diff
@@ -234,11 +248,20 @@ def rechunk(x, chunks="auto", threshold=None, block_size_limit=None):
         chunks, x.shape, limit=block_size_limit, dtype=x.dtype, previous_chunks=x.chunks
     )
 
+    # Now chunks are tuple of tuples
     if chunks == x.chunks:
         return x
     ndim = x.ndim
     if not len(chunks) == ndim:
         raise ValueError("Provided chunks are not consistent with shape")
+
+    if balance:
+        ideal_chunks = tuple(int(np.median(chunk)) for chunk in chunks)
+        even_chunks = tuple(
+            _even_chunksize(N, N // c) for N, c in zip(x.shape, ideal_chunks)
+        )
```

Comment:
There is a lot of logic here with median and with the `_even_chunksize` function.  Why is this necessary?  Would it be ok to look at the shape, the number of chunks, and then do the logic in your `_get_chunks` function?  Functions with lots of tricky logic in them make me nervous.  They often cause error reports in the future.  Is there something about the choices that this code makes that are important?

Reply:
Good point. The logic is convoluted. I think it's because I went from `n_chunks` to `balance`.

I think I see a way to simplify the logic in this PR. I'll check in after implementation.